upgrade.go

```diff
@@ -227,6 +227,11 @@ type Diff struct {
 
 	Diffs               []string
 	DeleteBeforeReplace bool
+
+	// May only be populated if there's a change.
```

Comment:
Curious about your vision for this. If this information is for diagnostic purposes only and not for making precise asserts on, we could get by with just exporting some text that renders the diff. If it's for making asserts on then I'd be curious what the `any` is - we're taking Diff and DiffConfig into this form, if that's so I suspect there's pretty mind-boggling accidental complexity in how the data is represented here. If we write a lot of asserts against this data, once the representation is cleaned up the assert code will have to change as well.

Not blocking on this just a discussion item.

Reply:
That's an interesting point. My use case was for diagnostic purposes only. As an author of a provider upgrade, when I have a test failure telling me there's a diff in `assumeRolPolicy`, I'd like to see what the diff is.